lsp-mode.el

```diff
@@ -1859,42 +1852,88 @@ WORKSPACE is the workspace that contains the progress token."
                  (const :tag "All Projects" :global))
   :package-version '(lsp-mode . "6.3"))
 
+(defvar-local lsp--diagnostics-modeline-string nil
+  "Value of current buffer diagnostics statistics.")
+(defvar lsp--diagnostics-modeline-wks->strings nil
+  "Plist of workspaces to their modeline strings.
+The `:global' workspace is global one.")
+
 (defun lsp--diagnostics-modeline-statistics ()
   "Calculate diagnostics statistics based on `lsp-diagnostics-modeline-scope'"
   (let ((diagnostics (cond
-                      ((equal :file lsp-diagnostics-modeline-scope)
-                       (lsp--get-buffer-diagnostics))
-                      (t (->> (eq :workspace lsp-diagnostics-modeline-scope)
-                              (lsp-diagnostics)
-                              (ht-values)
-                              (-flatten))))))
-    (->> diagnostics
-         (-group-by #'lsp:diagnostic-severity?)
-         (-sort (-lambda ((left) (right))
-                  (> right left)))
-         (-map (-juxt (-compose #'lsp--severity-code->severity #'cl-first)
-                      (-compose #'length #'cl-rest)))
-         (-map (-lambda ((code count))
-                 (propertize (format "%s" count)
-                             'face (cl-case code
-                                     ('error 'error)
-                                     ('warning 'warning)
-                                     ('info 'success)
-                                     ('hint 'success)))))
-         (s-join "/")
-         (format "%s"))))
+                       ((equal :file lsp-diagnostics-modeline-scope)
+                        (list (lsp--get-buffer-diagnostics)))
+                       (t (->> (eq :workspace lsp-diagnostics-modeline-scope)
+                               (lsp-diagnostics)
+                               (ht-values))))))
+    (->> (let ((stats (make-vector lsp/diagnostic-severity-max 0))
+               strs
+               (i 0))
+           (mapc (lambda (buf-diags)
+                   (mapc (lambda (diag)
+                           (-let [(&Diagnostic? :severity?) diag]
+                             (when severity?
+                               (cl-incf (aref stats severity?)))))
+                         buf-diags))
+                 diagnostics)
+           (while (< i lsp/diagnostic-severity-max)
```

Comment:
Seems like this can be rewritten using `seq-map-indexed`.

Reply:
I've tried that before, it either returns a list of results which we will need to filter out `0` value or we can let it `nconc` to `strs` and waste the returned list.